core/src/main/scala/org/bitcoins/core/gcs/GCS.scala

```diff
@@ -0,0 +1,226 @@
+package org.bitcoins.core.gcs
+
+import org.bitcoins.core.number.{UInt64, UInt8}
+import org.bitcoins.core.protocol.CompactSizeUInt
+import org.bouncycastle.crypto.macs.SipHash
+import org.bouncycastle.crypto.params.KeyParameter
+import scodec.bits.{BitVector, ByteVector}
+
+import scala.annotation.tailrec
+
+// TODO: Replace ByteVector with a type for keys
+/**
+  * Defines all functionality dealing with Golomb-Coded Sets
+  * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#GolombCoded_Sets]]
+  */
+object GCS {
```

Comment:
Just in case, I did a similar implementation in Scala:
- https://github.com/X9Developers/block-explorer/blob/develop/server/app/com/xsn/explorer/gcs/GolombEncoding.scala
- https://github.com/X9Developers/block-explorer/blob/develop/server/test/com/xsn/explorer/gcs/GolombEncodingSpec.scala

Reply:
Hey Alex! Sweet :) Any chance you've written the code to serialize script outputs (i.e. generate block filters from blocks)? I think I have all my GCS stuff in good shape but my serialization isn't passing tests :(

Reply:
No, I may end up working on that someday.

Is good to know I'm not the only one building those kind of pieces in Scala.

Reply:
Ah I've found it: https://github.com/X9Developers/block-explorer/commit/37cb7ceb81a9ce8d2a098b3b319b5fca616c2d2b

Any chance you've tested this logic against the bip 158 test vectors?
https://github.com/bitcoin/bips/blob/master/bip-0158/testnet-19.json

Reply:
That's not the serialization you are likely talking about, it basically collects all the addresses involved in a block and produces the filter, it does involve only the addresses from the `addresses` field from transaction inputs/outputs, I understand that serializing the whole script requires other details.

Reply:
Gotchya, well I just finally got serialization working today! So should you ever do it feel free to see how we did it :)

Reply:
Will do, thanks!

Reply:
Guess what? now I need to implement the filter for the script, I'll try to reuse this work.

Reply:
Good luck :) feel free to reach out with any questions! (Best way to reach us is probably through the Suredbits slack https://join.slack.com/t/suredbits/shared_invite/enQtNDEyMjY3MTg1MTg3LTYyYjkwOGUzMDQ4NDAwZjE1M2I3MmQyNWNlZjNlYjg4OGRjYTRjNWUwNjRjNjg4Y2NjZjAxYjU1N2JjMTU1YWM where there's a bitcoin-s channel)